### DISCUSSION TITLE
How to choose the value of the option `splitChunks.chunks`?

### DISCUSSION THREAD

**Discussion by nyngwang** [webpack/webpack#18040] [Q&A] [answered]
As title. From the documentation, quoted from [there](https://webpack.js.org/plugins/split-chunks-plugin/#splitchunkschunks):

> This indicates which chunks will be selected for optimization. When a string is provided, valid values are `all`, `async`, and `initial`. **Providing `all` can be particularly powerful**, because it means that chunks can be shared even between async and non-async chunks.

Without the knowledge of the implementation details of webpack, I will always choose `all` since it says it's _particularly powerful_. What are the considerations I should be aware of before adopting this always-`all`-strategy? (I did try to read the source but I still don't know the drawbacks of always choosing `all`)

**Comment by curly210102** [comment]
The 'initial', 'async', and 'all' values for the splitChunks.chunks option in webpack determine which chunks are targeted for optimization:

- 'initial': This value targets the entry chunk(s) of the application.
- 'async': It targets the dynamically imported (asynchronous) chunks.
- 'all': It includes all types of chunks, both initial and async.
Using a function provides even more fine-grained control over which chunks are selected for optimization. The function receives the chunk as an argument and should return a truthy value to indicate that the chunk should be optimized.

By choosing one of these values or defining a custom function, you can customize how webpack's SplitChunksPlugin optimizes and splits chunks in your application.

P.S. it's answered by AI chat bot, It's convenient to solve usage problems with ChatGPT or other AI bot.

**Reply by nyngwang** [reply]
But it didn't answer the drawback of always-`all`-strategy.

**Reply by alexander-akait** [reply, marked as answer]
We recommend using `all` by default, and you should stick to this strategy if you don't know. Sometimes it is useful for application performance to make more complex and detailed settings, you can want to reduce size of initial chunks (and load lazy async chunks), also you can want to merge all node_modules into one chunk and preload it before, you can want to merge all CSS chunks into one. 

It’s difficult to say which strategy will be best for you, because without knowing the structure of the application and the requirements for it, choosing the right settings is quite difficult. We have https://github.com/webpack-contrib/webpack-bundle-analyzer to help with it, there are also other tools which can show you and give advices, for example - https://github.com/statoscope/statoscope.

**Reply by kno-raziel** [reply]
> We recommend using `all` by default, and you should stick to this strategy if you don't know. Sometimes it is useful for application performance to make more complex and detailed settings, you can want to reduce size of initial chunks (and load lazy async chunks), also you can want to merge all node_modules into one chunk and preload it before, you can want to merge all CSS chunks into one.
> 
> It’s difficult to say which strategy will be best for you, because without knowing the structure of the application and the requirements for it, choosing the right settings is quite difficult. We have https://github.com/webpack-contrib/webpack-bundle-analyzer to help with it, there are also other tools which can show you and give advices, for example - https://github.com/statoscope/statoscope.

I'm wrong or this answer contradicts your previous answer here https://github.com/webpack/webpack/issues/16832#issuecomment-1475377014 😵‍💫

![image](https://github.com/user-attachments/assets/c9abcc69-0ce1-414d-944f-aa7a853aae85)


In that answer you added this link https://github.com/webpack/webpack/blob/main/lib/config/defaults.js#L1186 which is pointing to wrong line probably because file changed over the time, I guess you originally were pointing to this default https://github.com/webpack/webpack/blob/main/lib/config/defaults.js#L1504 which is the same as the one being used by [rspack](https://rspack.dev/plugins/webpack/split-chunks-plugin#default).

**Reply by alexander-akait** [reply]
Because this answer for a regular application, using “all” and module federation require detailed configurations, otherwise you will get broken runtime code, you still can use any value, but if you don’t understand why and how I recommend do not use “all” value for module federation. And yes we need to improve our docs and put a notice about it, and note related only for module federation
